sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1241,6 +1242,23 @@ object OptimizeRepartition extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove window partition if partition expressions are foldable
+ */
+object EliminateWindowPartitions extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
+    _.containsPattern(WINDOW), ruleId) {
+      case w @ Window(we, ps, _, _) if ps.exists(_.foldable) =>
+        val newWe = we.map {
+          _.transform {
+            case wsc @ WindowSpecDefinition(_ps, _, _) if _ps.exists(_.foldable) =>
+              wsc.copy(partitionSpec = _ps.filter(!_.foldable))
+          }.asInstanceOf[NamedExpression]
+        }
+        w.copy(windowExpressions = newWe, partitionSpec = ps.filter(!_.foldable))
```

Comment:
```
      case w @ Window(windowExpressions, partitionSpec, _, _) if partitionSpec.exists(_.foldable) =>
        val newWindowExpressions =
          windowExpressions.map(_.transformWithPruning(_.containsPattern(WINDOW_EXPRESSION)) {
            case wsd @ WindowSpecDefinition(ps, _, _) if ps.exists(_.foldable) =>
              wsd.copy(partitionSpec = ps.filter(!_.foldable))
          }.asInstanceOf[NamedExpression])
```

Reply:
Should this be done? `WINDOW_EXPRESSION` mismatch `WindowSpecDefinition`.
```
case w @ Window(we, ps, _, _) if ps.exists(_.foldable) =>
        val newWe = we.map(_.transformWithPruning(_.containsPattern(WINDOW_EXPRESSION)) {
          case _we @ WindowExpression(_, wsd @ WindowSpecDefinition(_ps, _, _))
            if _ps.exists(_.foldable) =>
            val newWsd = wsd.copy(partitionSpec = _ps.filter(!_.foldable))
            _we.copy(windowSpec = newWsd)
          }.asInstanceOf[NamedExpression])
        w.copy(windowExpressions = newWe, partitionSpec = ps.filter(!_.foldable))
```

Reply:
I haven't actually built it, just for reference. You should ensure that it is correct.

Could we make the variable name more readable? e.g. `we` => `windowExprs` or `windowExpressions` and others.

Reply:
I adjusted the above code
```
      case w @ Window(windowExprs, partitionSpec, _, _) if partitionSpec.exists(_.foldable) =>
        val newWe = windowExprs.map(_.transformWithPruning(_.containsPattern(WINDOW_EXPRESSION)) {
          case windowExpr @ WindowExpression(_, wsd @ WindowSpecDefinition(ps, _, _))
            if ps.exists(_.foldable) =>
            val newWsd = wsd.copy(partitionSpec = ps.filter(!_.foldable))
            windowExpr.copy(windowSpec = newWsd)
        }.asInstanceOf[NamedExpression])
        w.copy(windowExpressions = newWe, partitionSpec = partitionSpec.filter(!_.foldable))
```

Reply:
Updated.